examples/recurrent_attention_machine_translation.py

```diff
@@ -0,0 +1,683 @@
+        hit the word limit.
+
+# Data
+We use the machine translation dataset described in [2].
+To download the data run:
+    mkdir -p data/wmt16_mmt
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/training.tar.gz
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/validation.tar.gz
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/mmt16_task1_test.tar.gz
+    tar -xf training.tar.gz -C data/wmt16_mmt && rm training.tar.gz
+    tar -xf validation.tar.gz -C data/wmt16_mmt && rm validation.tar.gz
+    tar -xf mmt16_task1_test.tar.gz -C data/wmt16_mmt && rm mmt16_task1_test.tar.gz
+
+# References
+[1] Neural Machine Translation by Jointly Learning to Align and Translate
+    https://arxiv.org/abs/1409.0473
+[2] Multi30K: Multilingual English-German Image Descriptions
+    https://arxiv.org/abs/1605.00459
+    (http://www.statmt.org/wmt16/multimodal-task.html)
+
+# Differences between this implementation and [1]
+- A different/older dataset (wmt14) is used in [1].
+- The Tokenization here is similar but not identical to [1].
+- Initialisation of weights are not identical here and [1].
+- Normalisation of gradients is done when L2-norm > 1 in [1].
+- In the detailed description of the architecture in [1] it is stated:
+    "From here on, we omit all bias terms in order to increase
+    readability". It is thus not fully clear which linear transformations
+    also has bias terms, here all have.
+- TODO(2) Beam search is used in inference in [1]. This would be nice to
+    add to make example complete.
+- TODO(3) A "cascading architecture" is use in [1] by feeding not only
+    the GRU output to the readout layer but also the attention encoding.
+    This can be done by:
+    1) concatenating the attention encoding to the output of the
+        wrapped cell in the _RNNAttentionCell. However, then it must be
+        concatenated to the wrapped cell state as well which causes
+        cognitive overhead. This must be done because `state_size` and
+        `output_size` of an RNNCell must be the equal for masking to
+        work currently (this is a separate issue that can and should be
+        fixed though).
+    2) Adding "return_state_sequences" to RNN - this way concatenation
+        can be done externally and if also offers much more flexibility
+        for inspecting "what is attended" by the attention mechanism.
+- TODO(4) This implementation is inefficient in how attention is applied,
+    part of the computation can be done _once_ for the attended but is
+    now done at every timestep. It is kept this way here for readability.
+    Will make a separate PR to show alternative solution.
+(- There is no mention of Dropout or other regularisation methods in [1],
+    this could improve performance.)
+'''
+
+from __future__ import print_function
+
+import os
+
+from keras import backend as K
+from keras import (
+    initializers,
+    regularizers,
+    constraints)
+from keras.engine import (
```

Comment:
Style: don't import from `engine` (it's an internal factoring module). Instead do:

```python
import keras
from keras import layers
```

Then use e.g. `layers.Dense`